src/datasource.ts

```diff
@@ -0,0 +1,119 @@
+
+import { lastValueFrom } from 'rxjs';
+
+export class DataSource extends DataSourceApi<MyQuery, MyDataSourceOptions> {
+    url?: string;
+
+    constructor(instanceSettings: DataSourceInstanceSettings<MyDataSourceOptions>) {
+        super(instanceSettings);
+        this.url = instanceSettings.url;
+        this.annotations = {};
+    }
+
+    async query(options: DataQueryRequest<MyQuery>): Promise<DataQueryResponse> {
+        const { range } = options;
+        const from = range!.from.toISOString();
+        const to = range!.to.toISOString();
+
+        const promises = options.targets.map(async (target) => {
+            const response = this.doRequest(target, from, to);
+            const frame = new MutableDataFrame({
+                refId: target.refId,
+                fields: [
+                    { name: 'time', type: FieldType.time },
+                    { name: 'timeEnd', type: FieldType.string },
+                    { name: 'text', type: FieldType.string },
+                    { name: 'title', type: FieldType.string },
+                    { name: 'id', type: FieldType.string },
+                ],
+            });
+            await response
+                .then((response) => {
+                    response.data.incidents.forEach((incident: Incident) => {
+                        const timestamp: Date = new Date(incident.created_at);
+                        const timestamp_end: Date = new Date(incident.resolved_at);
+                        const createdAt = timestamp.getTime()
+                        const resolvedAt = timestamp_end.getTime();
+                        const title = incident.title;
+                        const text = incident.description;
+                        const id = incident.incident_key;
+
+                        frame.appendRow([createdAt, resolvedAt, title, text, id]);
+                    })
+                });
+            return frame;
+        });
+
+        return Promise.all(promises).then((data) => ({ data: data }));
+    }
+
+    async doRequest(query: MyQuery, from: string, to: string): Promise<FetchResponse<ListIncidentsResponse>> {
+        const routePath = '/pagerduty';
+
+        const result = getBackendSrv().fetch<ListIncidentsResponse>({
+            method: "GET",
+            url: this.url + routePath + "/incidents",
+            params: {
+                "service_ids[]": query.serviceId,
+                since: from,
+                until: to
+            },
+            responseType: "json"
+        })
+
+        return lastValueFrom(result);
+    }
+
+    async doRequestTest(): Promise<FetchResponse<ListIncidentsResponse>> {
+        const routePath = '/pagerduty';
+
+        const result = getBackendSrv().fetch<ListIncidentsResponse>({
+            method: "GET",
+            url: this.url + routePath + "/incidents",
+            responseType: "json"
+        })
+
+
+        return lastValueFrom(result);
+    }
+
+
+    async testDatasource() {
```

Comment:
Is this used?

Reply:
It is but it shouldn't be, I'll try to improved the datasource configuration validation